### ISSUE TITLE
Always falls back to 'ServerSentEvents'

### ISSUE THREAD

**Issue by dogwith1eye** [aspnet/SignalR#2458]
I'm a little puzzled why i can't get a websocket connection to work as the connection always falls back to 'ServerSentEvents' after printing the following error in the console.

```
WebSocket connection to 'wss://localhost:8001/hubs/notify?id=uNXrY7kf44Uj3f_MsqjSNg' failed: Error during WebSocket handshake: Unexpected response code: 500
(anonymous) @ WebSocketTransport.js:35

Utils.js:148 Error: Failed to start the transport 'WebSockets': undefined
```

I'm debugging in Visual Studio on Kestret which should support web sockets. Here is how i'm building the connection.

```
var notifyHubConnection = new HubConnectionBuilder()
                .withUrl("/hubs/notify")
                .configureLogging(LogLevel.Debug)
                .build();

```

**Comment by BrennanConroy** [comment]
You're using node I'm guessing, you probably need to polyfill WebSockets.

**Comment by dogwith1eye** [comment]
Node is being used by the AspNet JavaScript Services for HMR support, but the backend is AspNet. It actually worked fine when i was using the old signalr-client on .Net 2.0, but with 2.1 out now i thought it was time to rewrite it against the release version.

**Comment by anurse** [comment]
If I remember correctly, JavaScript Services does server-side rendering in NodeJS. Is that where this error is ocurring? I'm not sure WebSockets will work in that scenario. Are you positive that the old client was using WebSockets? It may have just been falling back to Server-Sent Events as well.

@SteveSandersonMS might have some more insight as he's more familiar with the JavaScript Services side of things. Does that environment support WebSockets?

**Comment by dogwith1eye** [comment]
Yeah it's definitely something with JavaScript services. I ran the app in the production container and it works fine. When I switch ASPNETCORE_ENVIRONMENT from Development to Production it works fine in the development container as well.

It well maybe that i simply wasn't aware that the app wasn't using websockets in development before, as SignalR never logged an error message when it falls back to 'ServerSentEvents' as it does now.

**Comment by SteveSandersonMS** [comment]
Does commenting out the "use hot module replacement" line from your app startup code also make it work when ASPNETCORE_ENVIRONMENT is Development?

If so it sounds like you've found some conflict between HMR and SignalR, though it's surprising as you'd be the first to report this.

**Comment by dogwith1eye** [comment]
Nope, simply commenting out //HotModuleReplacement = true does not fix the problem. I have to change the ASPNETCORE_ENVIRONMENT in the docker-compose.override.yml file to Production in order to avoid the problem, which of course turns off JavaScriptServices.

**Comment by anurse** [comment]
@SteveSandersonMS Is `WebSocket` available in the server-side rendering?

**Comment by SteveSandersonMS** [comment]
@anurse For the part of the execution that's taking place within Node, it's a standard Node environment so you have whatever intrinsics are available in the version of Node you're using.

I don't think current versions of Node have `WebSocket` by default. You're expected to use an NPM package like https://www.npmjs.com/package/ws

**Comment by anurse** [comment]
After chatting with @SteveSandersonMS to get some background on JavaScriptServices, it seems like what's likely happening here is that you have SignalR code running in your "pre-rendering" function. If this is the case, you will need to either:

1. Just ignore the warnings, assuming everything is working.
1. Disable both WebSockets and Server-Sent Events (see the upcoming [Configuration](https://github.com/aspnet/Docs/blob/master/aspnetcore/signalr/configuration.md) docs)
1. Install a WebSocket and/or Server-Sent Events Polyfill for Node JS

We're looking at making it possible to run the JS client in Node unmodified in 2.2, but in 2.1 you have to have the necessary polyfills installed.

**Comment by BrennanConroy** [comment]
Take a look at one of our Node samples
https://github.com/aspnet/SignalR-samples/blob/8c791ce3b4872dd03a864591a8231a4fc237896a/StockTickR/NodeClient/app.js#L3-L4

**Comment by hubaozhang2017** [comment]
I am using angular 6. I installed ws: ^6.0.0.  The error msg is still there.

**Comment by anurse** [comment]
Did you register it as the global WebSocket constructor? You need a line of code at the start of your app that does the following:

```javascript
global.WebSocket = require("ws");
```

**Comment by hubaozhang2017** [comment]
In angular code or typescript code, "require" is not used.

**Comment by anurse** [comment]
But in Angular you shouldn't need the polyfill as it's running in a browser. The polyfill is needed in Node JS environments because there isn't a global `WebSocket` constructor.

**Comment by hubaozhang2017** [comment]
Everything runs correctly except there is an error message in Debug console output, Error: Failed to start the transport 'WebSockets': undefined. 
In development environment, the angular code is launched by command "ng serve". It is run under a developmental server.
